### PR TITLE
Parameters: variants, not mux [v3]

### PR DESCRIPTION
**Pull request by clebergnu**
This PR is a collection of architectural and terminology arrangements.

It starts by not using the "mux" or "multiplexer" terminology in places where the (abstract) varianter interface is meant to be.  Then, it cleans up the `AvocadoParams` interfaces.  Finally, it moves the tests that are related to the mux code to the `yaml_to_mux` optional plugin, a primer that should be followed on other optional plugins.

---
Changes from v2 (#2343):
 - Two new commits fixing two issues related to "paused msg":
   * `avocado/utils/process.py: fix writing of GDB "commands" file`
   * `avocado/core/runner.py: remove ResultProxy reference`
 - Fixed the use of `paused` instead of `paused_msg` from `avocado.utils.process`
 - Rebased

Changes from v1 (#2336):
 - Added a missing `__init__.py` file on tests directory
 - Adapt new location of `yaml` files used on the tests
 - Rebased

### CODE REVIEW

**Comment by clebergnu** [comment]
@apahim FIY, Travis couldn't handle the 7 commits:

```
The job exceeded the maximum time limit for jobs, and has been terminated.
```

Things went OK until the 7th one though.

File: avocado/core/runner.py [hunk 1]
```diff
@@ -187,10 +187,9 @@ def _tick(self):
                 self.status = msg
                 self.job._result_events_dispatcher.map_method('test_progress',
                                                               False)
-                if msg['paused']:
-                    reason = msg['paused_msg']
-                    if reason:
-                        self.job.log.warning(reason)
+                paused_msg = msg['paused']
+                if paused_msg:
+                    self.job.log.warning(paused_msg)
```
Comment (ldoktor) [inline avocado/core/runner.py]: Yes, I definitely prefer a single variable to 2 variables, but would you please split the commits? It's not related to the other part where you change and limit the number of shared attributes.

File: avocado/plugins/variants.py [hunk 2]
```diff
@@ -67,7 +67,7 @@ def configure(self, parser):
                             "Shows the node content (variables)")
         env_parser = parser.add_argument_group("environment view options")
         env_parser.add_argument('-d', '--debug', action='store_true',
-                                dest="mux_debug", default=False,
+                                dest="debug", default=False,
```
Comment (ldoktor) [inline avocado/plugins/variants.py]: This debug is really specific option. I think `variant_debug` or `varianter_debug` would be more appropriate to prevent possible clash of this argument with general purpose `--debug` argument.

File: avocado/core/runner.py [hunk 3]
```diff
@@ -513,12 +513,12 @@ def _template_to_factory(template, variant):
...
         :type template: tuple
         :param variant: variant to be applied, usually containing
-                        the keys: mux_path, variant and variant_id
+                        the keys: paths, variant and variant_id
         :type variant: dict
         :return: tuple(new_test_factory, applied_variant)
         """
         var = variant.get("variant")
-        mux_path = variant.get("mux_path")
+        paths = variant.get("paths")
```
Comment (ldoktor) [inline avocado/core/runner.py]: This breaks the backward compatibility. Please check for `mux_path` presence at least until 52LTS is deprecated.

File: avocado/core/parameters.py [hunk 4]
```diff
@@ -44,7 +44,7 @@ class AvocadoParams(object):
     This is subject of change (separate file, perhaps)
     """
 
-    def __init__(self, leaves, test_id, paths, logger_name=None):
+    def __init__(self, leaves, paths, logger_name=None):
```
Comment (ldoktor) [inline avocado/core/parameters.py]: Yep, I hesitated to remove it to avoid breaking compat with pickle, but with 52 released we can do this sort of changes.

File: optional_plugins/varianter_yaml_to_mux/avocado_varianter_yaml_to_mux/__init__.py [hunk 5]
```diff
@@ -342,8 +335,6 @@ def configure(self, parser):
         """
         Configures "run" and "variants" subparsers
         """
-        if not MULTIPLEX_CAPABLE:
-            return
```
Comment (ldoktor) [inline optional_plugins/varianter_yaml_to_mux/avocado_varianter_yaml_to_mux/__init__.py]: Would you please split this little modification to a separate commit?

File: optional_plugins/varianter_yaml_to_mux/tests/test_multiplex.py [hunk 6]
```diff
@@ -10,17 +10,16 @@
...
-
-DEBUG_OUT = """Variant mint-debug-amd-virtio-07c6:    amd@selftests/.data/mux-environment.yaml, virtio@selftests/.data/mux-environment.yaml, mint@selftests/.data/mux-environment.yaml, debug@selftests/.data/mux-environment.yaml
-    /distro/mint:init         => systemv@selftests/.data/mux-environment.yaml:/distro/mint
-    /env/debug:opt_CFLAGS     => -O0 -g@selftests/.data/mux-environment.yaml:/env/debug
-    /hw/cpu/amd:cpu_CFLAGS    => -march=athlon64@selftests/.data/mux-environment.yaml:/hw/cpu/amd
-    /hw/cpu/amd:joinlist      => ['first_item']@selftests/.data/mux-selftest.yaml:/hw/cpu + ['second', 'third']@selftests/.data/mux-selftest.yaml:/hw/cpu/amd
-    /hw/disk/virtio:disk_type => virtio@selftests/.data/mux-environment.yaml:/hw/disk/virtio
-    /hw/disk:corruptlist      => nonlist@selftests/.data/mux-selftest.yaml:/hw/disk
-    /hw:corruptlist           => ['upper_node_list']@selftests/.data/mux-selftest.yaml:/hw
+AVOCADO = os.environ.get("UNITTEST_AVOCADO_CMD", "../scripts/avocado")
```
Comment (ldoktor) [inline optional_plugins/varianter_yaml_to_mux/tests/test_multiplex.py]: This will break the usage of `UNITTEST_AVOCADO_CMD` because one hardcodes it to `./scripts/avocado` environment but here you require different level. I think better solution would be `os.chdir` at the beginning of each test (using AVOCADO).

File: selftests/run [hunk 7]
```diff
@@ -26,6 +35,11 @@ def test_suite():
     for section in ('unit', 'functional', 'doc'):
         suite.addTests(loader.discover(start_dir=os.path.join(selftests_dir, section),
                                        top_level_dir=basedir))
+    # when other plugins get their own tests, make this a generic loop
+    if plugin_available('avocado-framework-plugin-varianter-yaml-to-mux'):
+        path = os.path.join(basedir, 'optional_plugins', 'varianter_yaml_to_mux')
+        suite.addTests(loader.discover(start_dir='tests',
+                                       top_level_dir=path))
```
Comment (ldoktor) [inline selftests/run]: This does not cover the `PARALLEL` or `AVOCADO` check.

Comment (ldoktor) [reply selftests/run]: The more I think about this I'd prefer to put optional_plugins tests into `selftests/optional_plugins/` directory. It won't be perfect from separation point of view, but it'd be way simpler to maintain and use, besides I don't think we could expect many optional-plugin selftests and if so the optional-plugin would probably be so complex that at that point it'd already be a separate entity like `avocado-virt`. What do you think?

File: avocado/core/runner.py (hunk 1 above)
Comment (clebergnu) [reply avocado/core/runner.py]: I see this change as a cohesive unit.  The goal is to reduce the number of needed attributes, and dropping `paused_msg` is just that.  It's not because the code is in some other file that it's *not* related to this change.

File: avocado/plugins/variants.py (hunk 2 above)
Comment (clebergnu) [reply avocado/plugins/variants.py]: I thought about that as well, but this is limited to the `variants` command so it seems OK to not need a prefix.  All other command line options are *not* using a prefix, so only this one using it would actually look odd.  What do you think?

File: avocado/core/parameters.py (hunk 4 above)
Comment (clebergnu) [reply avocado/core/parameters.py]: Yep.  I also forgot to remove the docstring.  Removing it in v+1.

File: selftests/run (hunk 7 above)
Comment (clebergnu) [reply selftests/run]: And the more I look at our custom shell/Python scripts to run our own tests, the more I think we are rewriting (non-optimally) what Avocado is meant to be doing.  Running tests in parallel is one thing.  Finding tests across directories is another one.

Anyway, I do think we should attempt to make plugins really that: separated and optional, including tests.  I'll address the other use cases such as PARALLEL.

**Comment by clebergnu** [comment]
@ldoktor I don't think I properly understand what you mean.  Do you mean that the last commit was hard to read and understand?

File: avocado/core/runner.py (hunk 3 above)
Comment (clebergnu) [reply avocado/core/runner.py]: You're missing the point regarding LTS.  This branch (master) does *not* have to keep compatibility with the LTS branch (52lts), not even with itself in core modules (as in previous versions).

What we won't do is break compatibility *within* the LTS branch (52lts) and with the test APIs in the master branch (so user's tests won't break on newer versions 57.0, 58.0, etc).

File: avocado/core/runner.py (hunk 1 above)
Comment (ldoktor) [reply avocado/core/runner.py]: My motivation was to separate the "facelift" in case some plugins require the removed attributes. But sure, it's not strong opinion.

File: avocado/plugins/variants.py (hunk 2 above)
Comment (ldoktor) [reply avocado/plugins/variants.py]: Not really, if you introduce subparser "debug", you make it impossible for main parser to define it as well. And it's even worse because the "debug" param is used in standard execution as well so if theoretically someone introduced "run" plugin with "debug" argument, it'd also start using TreeNodeDebug instead of TreeNode...

File: avocado/core/runner.py (hunk 3 above)
Comment (ldoktor) [reply avocado/core/runner.py]: The way I understand LTS is that users should be using this and also be able to experiment with master as that is to become next LTS. Then when new LTS is released they should seamlessly migrate there. So master does not need to read 32lts results, but it should be able to read 52lts results as the next LTS should be able to view previous LTS versions and that is how 52lts worked (it's able to read 32 results).

File: selftests/run (hunk 7 above)
Comment (ldoktor) [reply selftests/run]: yep, when Avocado has support to run things in parallel, we could simplify our execution, but until then parallel is important to me... (It's really fast with 8 cores)

File: avocado/core/runner.py (hunk 3 above)
Comment (apahim) [reply avocado/core/runner.py]: @clebergnu I see @ldoktor point here, it's related to the jobdata backward compatibility. We are storing the serialized variants and changing the names will avoid a further version of replaying the job created by a previous version. We have to agree on an approach for that, because last LTS was very painful to make compatible with the LTS before that. We should either have some migration tool, tracking and taking care of the compatibility breakages, or keep the compatibility code from one LTS all the way to the other. I'd prefer the former (and I just saw the card requesting that... +1 for it) and that why I accepted this code breaking the compatibility.

File: avocado/plugins/variants.py (hunk 2 above)
Comment (apahim) [reply avocado/plugins/variants.py]: Ok, I see @ldoktor point here. While harmless currently, it can lead us to future issues, so let's just use a less break-prone naming. Please have a look at https://github.com/avocado-framework/avocado/pull/2360

File: selftests/run (hunk 7 above)
Comment (clebergnu) [reply selftests/run]: Then please take a look at this:

https://github.com/avocado-framework/avocado/pull/2358

File: avocado/core/runner.py (hunk 3 above)
Comment (clebergnu) [reply avocado/core/runner.py]: We did go through extra steps to ensure that LTS data was readable by LTS+1.  But, AFAICT, it's not something we promised upfront in the definition of what LTS is.  I can go through the RFC again, but I'm pretty sure the extra 6 months of overlapping maintenance is exactly for that reason: users can evaluate the incompatibilities, work around them and then safely migrate to the new version.

What I don't want here, is to hold back further progress for the sake of "almost eternal" compatibility worries.

File: optional_plugins/varianter_yaml_to_mux/avocado_varianter_yaml_to_mux/__init__.py (hunk 5 above)
Comment (clebergnu) [reply optional_plugins/varianter_yaml_to_mux/avocado_varianter_yaml_to_mux/__init__.py]: This "little" modification means that all the tests in the current location will need an alternative check, that is, an "is plugin avaialable?" check.  I found it counter intuitive, but I've done so you can see what I mean clearly.
